src/corehost/cli/nethost/nethost.h

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef __NETHOST_H__
+#define __NETHOST_H__
+
+#include <cstddef>
+
+#if defined(_WIN32)
+    #ifdef NETHOST_EXPORT
+        #define NETHOST_API __declspec(dllexport)
+    #else
+        #define NETHOST_API __declspec(dllimport)
+    #endif
+
+    #define NETHOST_CALLTYPE __stdcall
+    #ifdef _WCHAR_T_DEFINED
+        using char_t = wchar_t;
+    #else
+        using char_t = unsigned short;
+    #endif
+#else
+    #ifdef NETHOST_EXPORT
+        #define NETHOST_API __attribute__((__visibility__("default")))
+    #else
+        #define NETHOST_API
+    #endif
+
+    #define NETHOST_CALLTYPE
+    using char_t = char;
+#endif
+
+//
+// Get the path to the hostfxr library
+//
+// Parameters:
+//   result_buffer
+//     Buffer that will be populated with the hostfxr path, including a null terminator.
+//
+//   buffer_size
+//     Size of result_buffer in char_t units
+//
+//   out_buffer_required_size
+//     Minimum required size in char_t units for a buffer to hold the hostfxr path
+//
+//   assembly_path
+//     Optional. Path to the compenent's assembly. Whether or not this is specified
+//     determines the behaviour for locating the hostfxr library.
+//     If nullptr, hostfxr is located using the enviroment variable or global registration
+//     If specified, hostfxr is located as if the assembly_path is the apphost
+//
+// Return value:
+//   0 on success, otherwise failure
+//   0x80008098 - result_buffer is too small (HostApiBufferTooSmall)
+//
+// Remarks:
+//   The full search for the hostfxr library is done on every call. To minimize the need
+//   to call this function multiple times, pass a large result_buffer (e.g. PATH_MAX).
+//
+extern "C" NETHOST_API int NETHOST_CALLTYPE nethost_get_hostfxr_path(
```

Comment:
Is it expected that native hosters carry this dll with them rather than it being app local? Ex if I have a global module, and I wanted to find hostfxr, I would need to have nethost.dll sit side by side with my global module?

Reply:
Yes, the expectation is that a native host application will ship this library as part of the application itself.

Reply:
I think I need a bit of clarification. The expectation is that a published app will contain the nethost.dll, like:

```
published/
    nethost.dll
    myapp.dll
```

Also, is the nethost.dll opt-in or opt-out?

Reply:
The `nethost.dll` is to be part of the native host - so in the case of ASP.NET it will be part of the ASP.NET IIS module.
* .NET Core app doesn't need it, since the `apphost` or muxer will find the `hostfxr` for it and if it needs the path to `hostfxr` at runtime it can simply ask for already loaded module.
* non .NET Core app which uses `hostfxr` to host .NET Core (this is what we call "native host") will include it in itself.

Basically the ASP.NET IIS module has the "locate hostfxr" code copied into it today, with this it would just carry it as a separate dll.

We don't have a delivery vehicle yet, but we're thinking native (C/C++) NuGet package and probably just raw `.zip` with the library and header file for all other cases (Linux/macOS, non-C/C++ languages).

Reply:
Our IIS module today installs via an msi, so I'd imagine we should probably include it SxS with the module. 

Once we get new bits from core-setup, I will start prototyping and see if all scenarios we used to support still work with `nethost.dll`.

Reply:
@jkotalik Just to clarify for me
> include it SxS with the module

This means that you would package the `nethost` with your module and install it into the same folder on the target machine... right?

Re the compatibility - if that's not the case (that is you find a case where it works differently then your existing code), please let us know. It's using the same code as `apphost` and so if there's a problem in `nethost` it's likely there is a problem in the other "hosts" as well.

Reply:
> This means that you would package the nethost with your module and install it into the same folder on the target machine... right?

Yes I believe that is the best course of action.

> Re the compatibility - if that's not the case (that is you find a case where it works differently then your existing code), please let us know. It's using the same code as apphost and so if there's a problem in nethost it's likely there is a problem in the other "hosts" as well.

I'm less concerned about that than some of the backups we did. For example, if all else fails with finding dotnet, we search in C:\PF\dotnet as a backup. So we will see!